firepit/clickhousestorage.py

```diff
@@ -0,0 +1,585 @@
+        cursor = self._execute(stmt)
+        if 'x_contained_by_ref' in columns:
+            #self._execute(f'CREATE INDEX {self.db_schema_prefix}"{tablename}_obs" ON "{tablename}" ("x_contained_by_ref");', cursor)
+            pass
+        self.connection.commit()
+        cursor.close()
+
+    def _add_column(self, tablename, prop_name, prop_type):
+        stmt = f'ALTER TABLE {self.db_schema_prefix}"{tablename}" ADD COLUMN "{prop_name}" {prop_type};'
+        logger.debug('new_property: "%s"', stmt)
+        cursor = self._execute(stmt)
+        cursor.close()
+
+    def _create_empty_view(self, viewname, cursor):
+        cursor.execute(f'CREATE VIEW {self.db_schema_prefix}"{viewname}" AS SELECT NULL as type WHERE 1<>1;')
+
+    def _create_view(self, viewname, select, sco_type, deps=None, cursor=None):
+        """Overrides parent"""
+        validate_name(viewname)
+        if not cursor:
+            cursor = self.connection.cursor()
+        if not deps:
+            deps = []
+        elif viewname in deps:
+            # Get the query that makes up the current view
+            slct = self._get_view_def(viewname)
+            self._execute(f'DROP VIEW IF EXISTS {self.db_schema_prefix}"{viewname}"', cursor)
+            # Swap out the viewname for its definition
+            select = re.sub(f'"{viewname}"', f'({slct}) AS tmp', select)
+        try:
+            cursor = self._execute(f'CREATE OR REPLACE VIEW {self.db_schema_prefix}"{viewname}" AS {select}', cursor)
+        except Exception:
+            #Ignore failure to create View
+            #traceback.print_stack()
+            return self.connection.cursor()
+        self._new_name(cursor, viewname, sco_type)
+        return cursor;
+
+
+    def _new_name(self, cursor, name, sco_type):
+        stmt = (f'INSERT INTO {self.db_schema_prefix}"__symtable" (name, type)'
+                f' VALUES (\'{name}\', \'{sco_type}\');')
+        cursor.execute(stmt)
+
+    def _drop_name(self, cursor, name):
+        stmt = f'ALTER TABLE {self.db_schema_prefix}"__symtable" DELETE WHERE name = \'{name}\';'
+        cursor.execute(stmt)
+
+    def _get_view_def(self, viewname):
+        cursor = self._query("SELECT create_table_query"
+                             " FROM system.tables"
+                             " WHERE database = '%s'"
+                             " AND name = '%s'"%(self.session_id, viewname))
+        viewdef = cursor.fetchone()
+        stmt = viewdef['create_table_query'].rstrip(';')
+        # PostgreSQL will "expand" the original "*" to the columns
+        # that existed at that time.  We need to get the star back, to
+        # match SQLite3's behavior.
+        return re.sub(r'^.*?FROM', 'SELECT * FROM', stmt, 1, re.DOTALL)
+
+    def tables(self):
+        cursor = self._query("SELECT name"
+                             " FROM system.tables"
+                             " WHERE database = '%s'"%(self.session_id))
```

Comment:
Instead of string formatting, can you use parameter substitution?  See https://docs.python.org/3/library/sqlite3.html#sqlite3-placeholders

Reply:
done